Add Supplements spider + integrated crawling/postprocessing

heroheman opened the pull request:
This PR adds first-class support for “supplement” pages from the SCP Wiki and integrates them into the existing crawl + post-processing workflow.

- Adds a new Scrapy spider `scp_supplement` to crawl pages tagged supplement and export them to `scp_supplement.json`.
- Introduces the ScpSupplement item type.
- Extends post-processing with `run_postproc_supplement` to generate processed outputs under supplement:
- `content_supplement.json `(full content + history/source/images)
- `index.json` (metadata + content_file)
- Adds `parent_scp` (best-effort extracted from the link) and parent_tale (best-effort extracted from *-<number> patterns).
- Updates the Makefile so Supplements are included in `scp_crawl` and `scp_postprocess` (and still available via dedicated `supplement_*` targets).
- Updates README documentation to mention the new spider, output file, and processed output.

## How to test

- `make scp` (includes supplements crawl + postprocess)

Or individually:
- `make supplement_crawl`
- `make supplement_postprocess`
- `scrapy crawl scp_supplement -o data/scp_supplement.json`

This PR should fix #2

heroheman committed:
chore(.gitignore): add .DS_Store to ignore list

heroheman committed:
feat(spiders): add SCP Supplement spider and item

- Introduce ScpSupplement class for item representation
- Implement ScpSupplementSpider to crawl supplement pages
- Update makefile to include supplement in data targets

heroheman committed:
feat(postprocessing): add supplement processing command

- Implement run_postproc_supplement to process SCP supplement data
- Create necessary directories and handle data extraction
- Store processed supplements in JSON format for further use

heroheman committed:
docs(README): update supplement crawl instructions and content structure

- Added instructions for crawling pages tagged as 'supplement'
- Updated content structure to include multiple content types
- Clarified post-processing details for supplements

heroheman committed:
fix(spiders): refine tale parsing rules to exclude unwanted links

- Updated the LinkExtractor in ScpTaleSpider to deny links
  matching specific patterns, improving the relevance of parsed tales.

heroheman committed:
feat(ci): add GitHub Actions workflow for SCP crawling

- for reference

tedivm commented on .github/workflows/scp-items.yml:
```diff
@@ -0,0 +1,92 @@
+        with:
+          repository: heroheman/scp_crawler
+          ref: "main"
+          path: scp-crawler
```
The crawler is just this repo isn't?

heroheman replied:
I removed this step.

tedivm commented on .github/workflows/scp-items.yml:
```diff
@@ -0,0 +1,92 @@
+      - name: "Clone this Repository"
+        uses: actions/checkout@v6
+        with:
+          path: scp-api
```
scp-api is a different repo than this one. It looks like this was copy/pasted from the other repo here.

I love the idea of having this test in here, but lets make sure we're pulling the right repos.

heroheman replied:
Hi, yes it was, just for debugging purposes, as the workflow is in scp-api. I just did not wanted to switch between browser and ide all the time - will remove it later when home! 

I removed the cloning step.

tedivm commented on .github/workflows/scp-items.yml:
```diff
@@ -0,0 +1,92 @@
+
+        env:
+          GIT_USER: "SCP Bot"
+          GIT_EMAIL: "scp@tedivm.com"
```
Since this is just a test, and the scp-api has its own workflow, lets have it do everything but push.

tedivm commented on .github/workflows/scp-items.yml:
```diff
@@ -0,0 +1,92 @@
+    branches:
+      - main
+    paths:
+      - .github/workflows/scp-items.yml
```
Lets have this run on PRs too so we can see when people make changes that break things.

tedivm reviewed (changes requested):
Overall this looks really good, but there are some changes to be made to the github workflows.

heroheman committed:
chore(workflow): adjusted workflow after review

- This change prevents automatic pushing to the SCP API during CI
- it should run on pull requests to allow for manual review first
- remove clone step as it's unnecessary
- resolves #5

heroheman committed:
fix(spiders): update tale parsing rules to allow all links

- this fixes unintend removal of Linkextractor Rule

heroheman committed:
fix(spiders): enhance error handling in history lookup

- Add checks for empty responses and missing 'body' in JSON
- Log errors for various failure scenarios to improve debugging
- Ensure robust parsing of history HTML to prevent crashes

heroheman committed:
chore(workflow): enable workflow_dispatch for SCP crawling

- Allows manual triggering of the workflow
- Improves flexibility for testing and updates

heroheman committed:
fix(history): improve history processing and error handling

- Handle empty history cases by returning an empty list
- Support both dict and list formats for history input
- Safely parse date strings with error handling
- Sort revisions by date, ensuring robustness against missing values

heroheman committed:
fix(history): ensure history key always exists in items

- Use `get` method to safely access history in hubs and items
- Prevent potential KeyError by ensuring history key is present
